catalog/rocks/catalog.go

```diff
@@ -0,0 +1,338 @@
+package rocks
+
+import (
+	"context"
+	"time"
+)
+
+// Basic Types
+
+// DiffType represents a changed state for a given entry (added, removed, changed, conflict)
+type DiffType uint8
+
+const (
+	DiffTypeAdded DiffType = iota
+	DiffTypeRemoved
+	DiffTypeChanged
+	DiffTypeConflict
+)
+
+// function/methods receiving the following basic types could assume they passed validation
+type (
+	// StorageNamespace is the URI to the storage location
+	StorageNamespace string
+
+	// RepositoryID is an identifier for a repo
+	RepositoryID string
+
+	// Path represents a logical path for an entry
+	Path string
+
+	// Ref could be a commit ID, a branch name, a Tag
+	Ref string
+
+	// TagID represents a named tag pointing at a commit
+	TagID string
+
+	// CommitID is a content addressable hash representing a Commit object
+	CommitID string
+
+	// BranchID is an identifier for a branch
+	BranchID string
+
+	// TreeID represents a snapshot of the tree, referenced by a commit
+	TreeID string
+
+	// StagingToken represents a namespace for writes to apply as uncommitted
+	StagingToken string
+
+	// CommonPrefix represents a path prefixing one or more Entry objects
+	CommonPrefix string
+
+	// Metadata key/value strings to holds metadata information on entry and commit
+	Metadata map[string]string
+)
+
+// Repository represents repository metadata
+type Repository struct {
+	StorageNamespace StorageNamespace
+	CreationDate     time.Time
+	DefaultBranchID  BranchID
+}
+
+type RepositoryRecord struct {
+	RepositoryID RepositoryID
+	*Repository
+}
+
+// Entry represents metadata or a given object (modified date, physical address, etc)
+type Entry struct {
+	LastModified time.Time
+	Address      string
+	Metadata     Metadata
+	ETag         string
+}
+
+// EntryRecord holds Path with the associated Entry information
+type EntryRecord struct {
+	Path Path
+	*Entry
+}
+
+// Commit represents commit metadata (author, time, tree ID)
+type Commit struct {
+	Committer    string
+	Message      string
+	TreeID       TreeID
+	CreationDate time.Time
+	Parents      []CommitID
+	Metadata     map[string]string
+}
+
+// CommitRecords holds CommitID with the associated Commit data
+type CommitRecord struct {
+	CommitID CommitID
+	*Commit
+}
+
+// Branch is a pointer to a commit
+type Branch struct {
+	CommitID CommitID
+	// nolint: structcheck, unused
+	stagingToken StagingToken
+}
+
+// BranchRecord holds BranchID with the associated Branch data
+type BranchRecord struct {
+	BranchID BranchID
+	*Branch
+}
+
+// Listing represents either an entry or a CommonPrefix
+type Listing struct {
+	CommonPrefix
+	*Entry
+}
+
+// Diff represents a change in path
+type Diff struct {
+	Path Path
+	Type DiffType
+}
+
+// Interfaces
+type Catalog interface {
+	// GetEntry returns entry from repository / reference by path, nil entry is a valid value for tombstone
+	// returns error if entry does not exist
```

Comment:
Let's specify which error (we had weirdness on this in the MVCC implementation).